File: csharp/Resources/NotificationChannelsResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Environment notification channel Wave 2 operations
    /// </summary>
    public class NotificationChannelsResource
    {
        private readonly AuthdogClient _client;

        public NotificationChannelsResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<JObject> ListAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/notification-channels");

        public JObject List(string tenantId, string environmentId) =>
            ListAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> CreateAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/notification-channels",
                body);

        public JObject Create(string tenantId, string environmentId, object body) =>
            CreateAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> UpdateAsync(string tenantId, string environmentId, string channelId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Put,
                $"{AuthdogClient.Env(tenantId, environmentId)}/notification-channels/{channelId}",
                body);

        public JObject Update(string tenantId, string environmentId, string channelId, object body) =>
            UpdateAsync(tenantId, environmentId, channelId, body).GetAwaiter().GetResult();

        public Task<JObject> DeleteAsync(string tenantId, string environmentId, string channelId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"{AuthdogClient.Env(tenantId, environmentId)}/notification-channels/{channelId}");

        public JObject Delete(string tenantId, string environmentId, string channelId) =>
            DeleteAsync(tenantId, environmentId, channelId).GetAwaiter().GetResult();

        public Task<JObject> TestAsync(
            string tenantId,
            string environmentId,
            string channelId,
            object? body = null) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/notification-channels/{channelId}/test",
                body);

        public JObject Test(string tenantId, string environmentId, string channelId, object? body = null) =>
            TestAsync(tenantId, environmentId, channelId, body).GetAwaiter().GetResult();
    }
}

File: java/src/main/java/com/authdog/resources/NotificationChannelsResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * Environment notification channel operations.
 */
public final class NotificationChannelsResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create a notification channels helper.
     * @param clientParam Authdog client
     */
    public NotificationChannelsResource(
            final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * List notification channels.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return channels envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode list(final String tenantId,
                         final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/notification-channels",
                null, null, JsonNode.class);
    }

    /**
     * Create a notification channel.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return created channel envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode create(final String tenantId,
                           final String environmentId,
                           final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/notification-channels",
                body, null, JsonNode.class);
    }

    /**
     * Update a notification channel.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param channelId channel ID
     * @param body request body
     * @return updated channel envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode update(final String tenantId,
                           final String environmentId,
                           final String channelId,
                           final Object body)
            throws AuthenticationException, ApiException {
        return client.request("PUT",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/notification-channels/" + channelId,
                body, null, JsonNode.class);
    }

    /**
     * Delete a notification channel.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param channelId channel ID
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode delete(final String tenantId,
                           final String environmentId,
                           final String channelId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/notification-channels/" + channelId,
                null, null, JsonNode.class);
    }

    /**
     * Test a notification channel.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param channelId channel ID
     * @return test envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode test(final String tenantId,
                         final String environmentId,
                         final String channelId)
            throws AuthenticationException, ApiException {
        return test(tenantId, environmentId, channelId, null);
    }

    /**
     * Test a notification channel with an optional body.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param channelId channel ID
     * @param body optional request body
     * @return test envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode test(final String tenantId,
                         final String environmentId,
                         final String channelId,
                         final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/notification-channels/" + channelId
                        + "/test",
                body, null, JsonNode.class);
    }
}
